File: dotnet/Pyrolite/Pickle/UnpickleStack.cs
/* part of Pyrolite, by Irmen de Jong (irmen@razorvine.net) */

using System.Collections;

namespace Razorvine.Pickle
{

/// <summary>
/// Helper type that represents the unpickler working stack. 
/// </summary>
public class UnpickleStack {
	private ArrayList stack;
	public object MARKER;

	public UnpickleStack() {
		stack = new ArrayList();
		MARKER = new object(); // any new unique object
	}

	public void add(object o) {
		stack.Add(o);
	}

	public void add_mark() {
		stack.Add(MARKER);
	}

	public object pop() {
		int size = stack.Count;
		var result = this.stack[size - 1];
		this.stack.RemoveAt(size-1);
		return result;
	}

	public ArrayList pop_all_since_marker() {
		ArrayList result = new ArrayList();
		object o = pop();
		while (o != this.MARKER) {
			result.Add(o);
			o = pop();
		}
		result.TrimToSize();
		result.Reverse();
		return result;
	}

	public object peek() {
		return this.stack[this.stack.Count-1];
	}

	public void trim() {
		this.stack.TrimToSize();
	}

	public int size() {
		return this.stack.Count;
	}

	public void clear() {
		this.stack.Clear();
		this.stack.TrimToSize();
	}
}

}

File: java/src/main/java/net/razorvine/pickle/UnpickleStack.java
package net.razorvine.pickle;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * Helper type that represents the unpickler working stack.
 * 
 * @author Irmen de Jong (irmen@razorvine.net)
 */
public class UnpickleStack implements Serializable {
	private static final long serialVersionUID = 6148106506441423350L;
	private ArrayList<Object> stack;
	protected Object MARKER;

	public UnpickleStack() {
		stack = new ArrayList<Object>();
		MARKER = new Object(); // any new unique object
	}

	public void add(Object o) {
		this.stack.add(o);
	}

	public void add_mark() {
		this.stack.add(this.MARKER);
	}

	public Object pop() {
		int size = this.stack.size();
		Object result = this.stack.get(size - 1);
		this.stack.remove(size - 1);
		return result;
	}

	public List<Object> pop_all_since_marker() {
		ArrayList<Object> result = new ArrayList<Object>();
		Object o = pop();
		while (o != this.MARKER) {
			result.add(o);
			o = pop();
		}
		result.trimToSize();
		Collections.reverse(result);
		return result;
	}

	public Object peek() {
		return this.stack.get(this.stack.size() - 1);
	}

	public void trim() {
		this.stack.trimToSize();
	}

	public int size() {
		return this.stack.size();
	}

	public void clear() {
		this.stack.clear();
		this.stack.trimToSize();
	}
}
